src/message-stream.ts

```diff
@@ -268,65 +321,90 @@ export class MessageStream extends PassThrough {
     client.initialize();
     return client.subscriberStub as Promise<ClientStub>;
   }
+
   /**
    * Since we do not use the streams to ack/modAck messages, they will close
    * by themselves unless we periodically send empty messages.
    *
    * @private
    */
   private _keepAlive(): void {
-    this._streams.forEach((receivedStatus, stream) => {
-      // its possible that a status event fires off (signaling the rpc being
-      // closed) but the stream hasn't drained yet, writing to this stream will
-      // result in a `write after end` error
-      if (!receivedStatus) {
-        stream.write({});
+    this._streams.forEach(tracker => {
+      // It's possible that a status event fires off (signaling the rpc being
+      // closed) but the stream hasn't drained yet. Writing to such a stream will
+      // result in a `write after end` error.
+      if (!tracker.receivedStatus && tracker.stream) {
+        tracker.stream.write({});
       }
     });
   }
+
+  // Returns the number of tracked streams that contain an actual stream (good or not).
+  private _activeStreams(): number {
+    return this._streams.reduce((p, t) => (t.stream ? 1 : 0) + p, 0);
+  }
+
   /**
    * Once the stream has nothing left to read, we'll remove it and attempt to
    * refill our stream pool if needed.
    *
    * @private
    *
-   * @param {Duplex} stream The ended stream.
+   * @param {number} index The ended stream.
    * @param {object} status The stream status.
    */
-  private _onEnd(stream: PullStream, status: grpc.StatusObject): void {
-    this._removeStream(stream);
-
-    if (this._fillHandle) {
-      return;
-    }
-
-    if (this._retrier.retry(status)) {
-      const delay = this._retrier.createTimeout();
-      this._fillHandle = setTimeout(() => this._fillStreamPool(), delay);
-    } else if (!this._streams.size) {
+  private _onEnd(index: number, status: grpc.StatusObject): void {
+    this._removeStream(index);
+
+    if (PullRetry.retry(status)) {
+      this.emit(
+        'debug',
+        new Error(
+          `Subscriber stream ${index} has ended with status ${status.code}; will be retried.`
+        )
+      );
+      if (PullRetry.resetFailures(status)) {
+        this._retrier.reset(this._streams[index]);
+      }
+      this._retrier.retryLater(this._streams[index], () => {
+        this._fillStreamPool();
```

Comment:
So we are still kind of "batching" the retries of streams given that we are using `_fillStreamPool`, is that right? So if two streams fail, we will retry both of them at whatever point the soonest one wants to be retried, if I'm reading this correctly. I rather we do the retries of individual streams independently.

Reply:
I would say that the appearance of batching is superficial at this point. I mostly just didn't want to completely restructure the code. This has changed so that _fillStreamPool just looks to see if any of the streams need replacing and does that. I guess in practice it could end up retrying something sooner than it's supposed to be retried, but all of the streams should be going to the same place anyway. (So if one succeeds, all will succeed.)

Reply:
I went ahead and split the retry up into its own method, so they're fully independent now.